bootstore/src/db/mod.rs

```diff
@@ -0,0 +1,230 @@
+
+//! Database layer for the bootstore
+
+mod macros;
+mod models;
+mod schema;
+
+use diesel::connection::SimpleConnection;
+use diesel::prelude::*;
+use diesel::SqliteConnection;
+use slog::Logger;
+use slog::{info, o};
+
+use crate::trust_quorum::SerializableShareDistribution;
+use models::EncryptedRootSecret;
+use models::KeyShareCommit;
+use models::KeySharePrepare;
+use models::Sha3_256Digest;
+use models::Share;
+use sha3::{Digest, Sha3_256};
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Failed to open db connection to {path}: {err}")]
+    DbOpen { path: String, err: ConnectionError },
+
+    #[error(transparent)]
+    Db(#[from] diesel::result::Error),
+
+    #[error(transparent)]
+    Json(#[from] serde_json::Error),
+
+    #[error("Share commit for {epoch} does not match prepare")]
+    CommitHashMismatch { epoch: i32 },
+}
+
+/// The ID of the database used to store blobs.
+///
+/// We separate them because they are encrypted and accessed differently.
+pub enum DbId {
+    /// Used pre-rack unlock: Contains key shares and membership data
+    TrustQuorum,
+
+    /// Used post-rack unlock: Contains information necessary for setting
+    /// up NTP.
+    NetworkConfig,
+}
+
+pub struct Db {
+    log: Logger,
+    conn: SqliteConnection,
+}
+
+impl Db {
+    pub fn open(log: Logger, path: &str) -> Result<Db, Error> {
+        let schema = include_str!("./schema.sql");
+        let log = log.new(o!(
+            "component" => "BootstoreDb"
+        ));
+        info!(log, "opening database {:?}", path);
+        let mut c = SqliteConnection::establish(path)
+            .map_err(|err| Error::DbOpen { path: path.into(), err })?;
+
+        // Enable foreign key processing, which is off by default. Without
+        // enabling this, there is no referential integrity check between
+        // primary and foreign keys in tables.
+        diesel::sql_query("PRAGMA foreign_keys = 'ON'").execute(&mut c)?;
+
+        // Enable the WAL.
+        diesel::sql_query("PRAGMA journal_mode = 'WAL'").execute(&mut c)?;
+
+        // Force overwriting with 0s on delete
+        diesel::sql_query("PRAGMA secure_delete = 'ON'").execute(&mut c)?;
+
+        // Sync to disk after every commit.
+        // DO NOT CHANGE THIS SETTING!
+        diesel::sql_query("PRAGMA synchronous = 'FULL'").execute(&mut c)?;
+
+        // Create tables
+        c.batch_execute(&schema)?;
+
+        Ok(Db { log, conn: c })
+    }
+
+    pub fn prepare_share(
+        &mut self,
+        epoch: i32,
+        share: SerializableShareDistribution,
+    ) -> Result<(), Error> {
+        use schema::key_share_prepares::dsl;
+        // We save the digest so we don't have to deserialize and recompute most of the time.
+        // We'd only want to do that for a consistency check occasionally.
+        let val = serde_json::to_string(&share)?;
+        let share_digest =
+            sprockets_common::Sha3_256Digest(Sha3_256::digest(&val).into())
```

Comment:
I think this is fine but just want to think out loud - we're hashing the `serde_json`'ified `share` here, but then inserting `share` directly into the `KeySharePrepare` that is then committed to the DB. Is there any chance we're sensitive to minor changes in serde_json that wouldn't affect correctness from a purely JSON point of view (reordering fields, adding or removing whitespace) by splitting things this way?

~~This might become clear from other parts of the PR I haven't read yet, but how does diesel serialize the `share` when inserting into the DB?~~ (Answered this question when reading `json_new_type`)

 I'm wondering if to check the digest, someone has to deserialize `share` then reserialize it with `serde_json` and hope that whatever version of `serde_json` they're using serialized it byte-for-byte identically to the version used to compute this digest. I think that is not necessary, right? The string we insert into the `share` column should be byte-for-byte identical to `val` in this function, assuming `serde_json::to_string()` is deterministic (which I think it should be for simple structs).

Reply:
That's a good question. The JSON will be serialized once when inserted into CockroachDB and once here. If serialization was not deterministic, we would have a problem. In fact, what we really want is "canonical" serialization, but I don't believe serde_json provides that. This is required by TUF as well fwiw, and other crypto projects. From what I can tell though this isn't a problem for us, unless the serialization format drastically changes, which while unlikely is a bit scare. We have the need to hash and sign all this data and much other data, so maybe JSON isn't the right choice for a format here.

Reply:
Ok, I used [BCS](https://github.com/diem/bcs) instead of `serde_json` as of https://github.com/oxidecomputer/omicron/pull/1656/commits/815f8e16ad77412fc23f1be6f257100912de1ad6

It's the only serde backed canonical serialization format I could find and seems perfectly suitable here. It's practically a drop in replacement.